### PR TITLE
Improve Authentication Options

### PR DESCRIPTION
**Pull request by eric-spitler**
Add or improve upon authentication methods for WMS/WCS/WFS/WMTS/TMS services

In some cases, username/password authentication was partially there - service-level attributes set but capabilities-level not set. In other cases, no authentication was capable.

Changes:
- For `*Service`, `*Capabilities`, `ContentMetadata` classes, allow `username` | `password` | `cert` | `verify` arguments to `__init__()`
- For uses of `openURL`, include additional keyword arguments for `username` | `password` | `cert` | `verify`
- For uses of `requests` library, include additional keyword arguments for `username` | `password` | `cert` | `verify` (uses `auth=(username, password)`)
- When creating other class instances, include current values for these arguments

Default behavior for these arguments in the `requests` and `openURL` implementation is to default to `None`, which for username/password authentication means "unauthenticated".

For certificate-based authentication, defaulting to `cert=None` means unauthenticated. In the case of 2-way certificate authentication, paths to a cert/key combination is used as the `cert=` argument and the path a CA bundle is used for the `verify=` argument. If `verify=True` (default), use system CA bundle.

### CODE REVIEW

**Comment by eric-spitler** [comment]
Should resolve https://github.com/geopython/OWSLib/issues/412

**Comment by eric-spitler** [comment]
Enables use of `cert` and `verify` for https://github.com/geopython/OWSLib/issues/489

**Comment by tomkralidis** [comment]
Thanks @eric-spitler .  I've asked @cehbrecht to review as well.  Can the changes incorporate CSW as well?  I'm also wondering whether we need a custom `owslib.util.auth` object which gets passed around instead of tacking on parameters across functions all the time.  Thoughts?

File: VERSION.txt [hunk 1]
```diff
@@ -1 +1 @@
-0.17.1
+0.17.2
```
Comment (cehbrecht) [inline VERSION.txt]: Change version with next release?

**Comment by cehbrecht** [comment]
Tests failures are not related to the PR.

**Comment by cehbrecht** [comment]
> I'm also wondering whether we need a custom owslib.util.auth object which gets passed around instead of tacking on parameters across functions all the time. Thoughts?

@eric-spitler @tomkralidis Would be nice to simplify this code :) But that is probably for another PR. Could we use a base OWS class with an internal `openURL` method which has all the auth parameters from the class instance. Maybe that would avoid some of the parameter passing.

File: VERSION.txt (hunk 1 above)
Comment (eric-spitler) [reply VERSION.txt]: Sure, I'll get that changed back in both files.

**Comment by eric-spitler** [comment]
> > I'm also wondering whether we need a custom owslib.util.auth object which gets passed around instead of tacking on parameters across functions all the time. Thoughts?
> 
> @eric-spitler @tomkralidis Would be nice to simplify this code :) But that is probably for another PR. Could we use a base OWS class with an internal `openURL` method which has all the auth parameters from the class instance. Maybe that would avoid some of the parameter passing.

Completely agree. There so many places where the same changes were made. I was mostly trying to retain existing structure/functionality so as not to impact existing testing and utilization, but a consolidated solution would be best.

Also, some of the changes here are things I've done on a private project with local modifications, specifically around WMS/WCS and cert/verify usage. In that case I have a real site I'm accessing that requires that form of authentication, but I'd have to think about how to properly test that w/o relying on someone else's site/data.

**Comment by eric-spitler** [comment]
Just a quick update, I'm working on a centralized authentication solution but have a couple test kinks to work out. Basically doing what you said with an `Authentication` class in `owslib.util` and updating existing uses of `username` and `password` to merge into the new class. That class then has `.openURL(*args, **kwargs)` and `.get(*args, **kwargs)` to handle authenticated requests.

**Comment by tomkralidis** [comment]
Thanks @eric-spitler.  Should we wait for that work to come through in lieu of this PR?

**Comment by eric-spitler** [comment]
Work got pretty busy lately so I don't know when I'll be able to finish the updates. I would say merge, but they'll make backwards compatibility messy. If you're okay waiting I'll keep going on it.

**Comment by tomkralidis** [comment]
Perhaps it’s better to wait rather than introduce then replace functionality.

**Comment by eric-spitler** [comment]
@tomkralidis @cehbrecht Updates are ready for your review.

File: owslib/csw.py [hunk 2]
```diff
@@ -658,7 +660,7 @@ def _invoke(self):
 
         if isinstance(self.request, six.string_types):  # GET KVP
             self.request = '%s%s' % (bind_url(request_url), self.request)
-            self.response = openURL(self.request, None, 'Get', username=self.username, password=self.password, timeout=self.timeout).read()
+            self.response = self.auth.openURL(self.request, None, 'Get', timeout=self.timeout).read()
```
Comment (tomkralidis) [inline owslib/csw.py]: Should `openURL` be a function of the new Authentication mechanism or should we do it the other way around (i.e. pass an auth object as a parameter to `openURL`)

File: owslib/csw.py [hunk 3]
```diff
@@ -35,7 +33,7 @@
 from owslib.dif import DIF
 from owslib.gm03 import GM03
 from owslib.namespaces import Namespaces
-from owslib.util import cleanup_namespaces, bind_url, add_namespaces, openURL
+from owslib.util import cleanup_namespaces, bind_url, add_namespaces, OrderedDict, Authentication
```
Comment (tomkralidis) [inline owslib/csw.py]: Why we adding `OrderedDict` here?

File: owslib/util.py [hunk 4]
```diff
@@ -17,6 +18,7 @@
 from owslib.etree import etree, ParseError
 from owslib.namespaces import Namespaces
 from six.moves.urllib.parse import urlsplit, urlencode, urlparse, parse_qs, urlunparse
+import traceback
```
Comment (tomkralidis) [inline owslib/util.py]: Why are we adding `traceback` here?

**Comment by tomkralidis** [comment]
thanks @eric-spitler for the impressive work here!  See my review/comments. In addition:

- can you fix merge conflicts?
- aside: I imagine given the nature of the enhancement it is difficut to provide tests against services requiring authentication.  It would be nice to have test services at some point to help test. This shouldn't stop this PR from being merged however once it is approved.

File: owslib/csw.py (hunk 3 above)
Comment (eric-spitler) [reply owslib/csw.py]: Not sure how that got added. I'll remove it.

File: owslib/util.py (hunk 4 above)
Comment (eric-spitler) [reply owslib/util.py]: Not sure how that got added. I'll remove it.

File: owslib/csw.py (hunk 2 above)
Comment (eric-spitler) [reply owslib/csw.py]: Its up to you. I was thinking make the auth passing transparent, but can do it either way.

Comment (tomkralidis) [reply owslib/csw.py]: @cehbrecht thoughts?

Comment (eric-spitler) [reply owslib/csw.py]: Implemented

**Comment by eric-spitler** [comment]
The tests had 500/502 responses from external locations. Is there a way to trigger a re-run or just a new commit?

**Pull request merged**